aggregator/src/aggregator.rs

```diff
@@ -1953,7 +1953,7 @@ impl VdafOps {
                 let accumulator = Arc::clone(&accumulator);
 
                 Box::pin(async move {
-                    let unwritable_reports = accumulator.flush_to_datastore(tx, &vdaf).await?;
+                    let unwritable_reports = accumulator.unwritable_reports(tx, &vdaf).await?;
```

Comment:
`unwritable_reports` (the variable) is used to determine if we should mark an existing report aggregation as `BatchCollected`.

With this change, we introduce `unwritable_reports` (the accumulator method) to compute the set of reports that would/will be unwritable, without changing the datastore. I think this works, but it's less than ideal for a few reasons: we now have to maintain the same logic in two separate functions (`unwritable_reports` & `flush_to_datastore`); and we have to maintain more code.

Instead, WDYT about dropping the `unwritable_reports` method; instead, move the preexisting call to `flush_to_datastore` down to where you moved it in this PR, but execute it before the `try_join!` rather than as part of the `try_join!` and update the report aggregations to `BatchCollected` based on those changes.

Reply:
Nice suggestion!

I was able to have `flush_to_datastore()` still execute concurrently with the other insertion functions, by using some `.map()` trickery on the return value.

This lowered my confidence in the accumulator unit test, so I moved it further out to work at the HTTP level, with the other aggregate init tests in `http_handlers.rs`.

I also was also able to negotiate with `rustfmt` to format this code (the problem is the first `try_join!` in the closure, temporarily removing the `!` allows rustfmt to format the whole block). Apologies that the diff is somewhat mangled now.

Reply:
Good gracious, my latest commit makes this block formattable, by declaring the futures outside of the `try_join!` statement and keeping their variable names short :shrug: